File: python/peer_review_assistant/preprocess/section_normalizer.py
"""Section heading normalization for structure check.

Resolves manuscript section headings to canonical names using journal-specific
aliases and position-based disambiguation rules.

Usage:
    from peer_review_assistant.preprocess.section_normalizer import normalize_sections

    result = normalize_sections(section_map, journal_profile)
    # result = {
    #     "sections": [...],        # original sections + canonical_name + classification
    #     "canonical_order": [...], # deduplicated canonical names in document order
    #     "classifications": {...}, # canonical_name -> "front_matter"|"main_text"|"back_matter"
    # }
"""


def normalize_sections(section_map, journal_profile=None):
    """Normalize section headings to canonical names using journal aliases & rules.

    Args:
        section_map: dict with "sections" list. Each section has "name", "heading",
                     "level", "parent_section", "start_paragraph", "end_paragraph".
        journal_profile: Optional dict with "manuscript_structure" containing
                         section_aliases, section_alias_rules, main_text_order,
                         front_matter_sections, back_matter_sections.

    Returns:
        dict with keys: sections (list of normalized section dicts),
        canonical_order (list of canonical_name strings in document order),
        classifications (dict mapping canonical_name to classification).
    """
    sections = section_map.get("sections", []) if section_map else []
    if not sections:
        return {"sections": [], "canonical_order": [], "classifications": {}}

    ms = (journal_profile or {}).get("manuscript_structure", {})
    aliases = ms.get("section_aliases", {}) or {}
    rules = ms.get("section_alias_rules", []) or []
    main_text_order = ms.get("main_text_order", []) or []
    front_matter = ms.get("front_matter_sections", []) or []
    back_matter = ms.get("back_matter_sections", []) or []

    # ── Step 1: Build alias lookup ──
    alias_to_canonical: dict[str, str] = {}
    if aliases:
        for canonical, alias_list in aliases.items():
            for alias in alias_list:
                alias_to_canonical[alias.lower().strip()] = canonical

    # ── Step 2: Parse rules into lookup ──
    rule_lookup: dict[str, list[tuple[str, str]]] = {}
    for rule in rules:
        key = (rule.get("alias") or "").lower().strip()
        canon = rule.get("canonical", "")
        cond = rule.get("condition", "")
        if key and canon:
            rule_lookup.setdefault(key, []).append((canon, cond))

    # ── Step 3: First pass — resolve canonical names via aliases ──
    normalized = []
    for sec in sections:
        heading = (sec.get("heading") or "").strip()
        heading_lower = heading.lower()
        canonical = None

        # 3a. Exact alias match
        if heading_lower in alias_to_canonical:
            canonical = alias_to_canonical[heading_lower]

        # 3b. Substring alias match (e.g. "2.1 Introduction")
        if canonical is None and heading_lower:
            best_len = 0
            for alias_lower, canon_name in alias_to_canonical.items():
                if alias_lower in heading_lower and len(alias_lower) > best_len:
                    canonical = canon_name
                    best_len = len(alias_lower)

        # 3c. Fall back to section_map's existing name
        if canonical is None:
            canonical = sec.get("name")

        normalized.append({
            **sec,
            "canonical_name": canonical,
        })

    # ── Step 4: Second pass — apply position-based rules ──
    for i, sec in enumerate(normalized):
        heading = (sec.get("heading") or "").strip()
        heading_lower = heading.lower()

        if heading_lower in rule_lookup:
            for canon, condition in rule_lookup[heading_lower]:
                if _condition_holds(condition, i, normalized):
                    sec["canonical_name"] = canon
                    break  # first matching rule wins

    # ── Step 5: Classify each section ──
    classifications: dict[str, str] = {}
    for sec in normalized:
        canon = sec["canonical_name"]
        if _is_front_matter(canon, front_matter):
            classifications[sec["name"]] = "front_matter"
        elif _is_back_matter(canon, back_matter):
            classifications[sec["name"]] = "back_matter"
        elif _is_main_text(canon, main_text_order):
            classifications[sec["name"]] = "main_text"
        else:
            classifications[sec["name"]] = "main_text"  # default

    # ── Step 6: Build canonical_order (deduplicated) ──
    canonical_order = []
    seen: set[str] = set()
    for sec in normalized:
        cn = sec["canonical_name"]
        if cn not in seen:
            canonical_order.append(cn)
            seen.add(cn)

    return {
        "sections": normalized,
        "canonical_order": canonical_order,
        "classifications": classifications,
    }


# ── Condition evaluators ─────────────────────────────────────────────────

def _condition_holds(condition: str, idx: int, sections: list[dict]) -> bool:
    """Evaluate a position condition against the current section's document context."""
    sec = sections[idx]

    if condition == "before_introduction":
        for prev in sections[:idx]:
            if prev.get("canonical_name") == "introduction":
                return False
        return True

    if condition == "inside_or_immediately_after_discussion":
        parent = sec.get("parent_section")
        if parent and _canonical_of_parent(parent, sections) == "discussion":
            return True
        if idx > 0 and sections[idx - 1].get("canonical_name") == "discussion":
            return True
        return False

    if condition == "inside_methods":
        parent = sec.get("parent_section")
        if parent and _canonical_of_parent(parent, sections) == "methods":
            return True
        return False

    return False


def _canonical_of_parent(parent_name: str, sections: list[dict]) -> str | None:
    """Find the canonical_name of a section with the given section_map name."""
    for sec in sections:
        if sec.get("name") == parent_name:
            return sec.get("canonical_name")
    return None


# ── Classification helpers ───────────────────────────────────────────────

def _normalize_for_classification(canonical: str) -> str:
    """Strip special suffixes from canonical names for classification matching."""
    for suffix in ("_subsection", "_or_conclusion"):
        if suffix in canonical:
            canonical = canonical.split(suffix)[0]
    return canonical.lower().strip()


def _is_front_matter(canonical: str, front_list: list[str]) -> bool:
    base = _normalize_for_classification(canonical)
    return base in [f.lower().strip() for f in front_list]


def _is_back_matter(canonical: str, back_list: list[str]) -> bool:
    base = _normalize_for_classification(canonical)
    return base in [b.lower().strip() for b in back_list]


def _is_main_text(canonical: str, main_text_order: list[str]) -> bool:
    base = _normalize_for_classification(canonical)
    return base in [m.lower().strip() for m in main_text_order]

File: rust/pra-cli/src/section_normalizer.rs
/// Section heading normalization for structure check.
///
/// Port of `python/peer_review_assistant/preprocess/section_normalizer.py`.
/// Resolves manuscript section headings to canonical names using journal-specific
/// aliases and position-based disambiguation rules.
use serde_json::Value;

/// Normalize section headings to canonical names using journal aliases & rules.
///
/// `section_map` is a JSON value with a `"sections"` list.
/// `journal_profile` is an optional JSON value with `manuscript_structure`.
///
/// Returns a JSON value with keys `sections`, `canonical_order`, `classifications`.
pub fn normalize_sections(section_map: &Value, journal_profile: Option<&Value>) -> Value {
    let sections = section_map
        .get("sections")
        .and_then(|s| s.as_array())
        .cloned()
        .unwrap_or_default();

    if sections.is_empty() {
        return serde_json::json!({
            "sections": [],
            "canonical_order": [],
            "classifications": {}
        });
    }

    let ms = journal_profile
        .and_then(|jp| jp.get("manuscript_structure"));

    // Step 1: Build alias_to_canonical lookup
    // aliases is Map<String, Value> mapping canonical name → array of alias strings
    let mut alias_to_canonical: Vec<(String, String)> = Vec::new();
    if let Some(aliases) = ms.and_then(|m| m.get("section_aliases")).and_then(|a| a.as_object()) {
        for (canonical, alias_list) in aliases {
            if let Some(aliases_arr) = alias_list.as_array() {
                for alias in aliases_arr {
                    if let Some(a) = alias.as_str() {
                        let key = a.to_lowercase().trim().to_string();
                        if !key.is_empty() {
                            alias_to_canonical.push((key, canonical.clone()));
                        }
                    }
                }
            }
        }
    }

    // Step 2: Parse rules into lookup
    let rules = ms.and_then(|m| m.get("section_alias_rules")).and_then(|r| r.as_array()).cloned().unwrap_or_default();
    let mut rule_lookup: Vec<(String, Vec<(String, String)>)> = Vec::new();
    for rule in &rules {
        let key = rule
            .get("alias")
            .and_then(|a| a.as_str())
            .map(|s| s.to_lowercase().trim().to_string())
            .unwrap_or_default();
        let canon = rule
            .get("canonical")
            .and_then(|c| c.as_str())
            .unwrap_or("");
        let condition = rule
            .get("condition")
            .and_then(|c| c.as_str())
            .unwrap_or("");
        if !key.is_empty() && !canon.is_empty() {
            let mut found = false;
            for (rk, entries) in rule_lookup.iter_mut() {
                if rk == &key {
                    entries.push((canon.to_string(), condition.to_string()));
                    found = true;
                    break;
                }
            }
            if !found {
                rule_lookup.push((key, vec![(canon.to_string(), condition.to_string())]));
            }
        }
    }

    // Extract config lists for classification
    let main_text_order: Vec<String> = ms.and_then(|m| m.get("main_text_order")).and_then(|a| a.as_array())
        .map(|arr| arr.iter().filter_map(|v| v.as_str().map(|s| s.to_lowercase().trim().to_string())).collect())
        .unwrap_or_default();
    let front_matter: Vec<String> = ms.and_then(|m| m.get("front_matter_sections")).and_then(|a| a.as_array())
        .map(|arr| arr.iter().filter_map(|v| v.as_str().map(|s| s.to_lowercase().trim().to_string())).collect())
        .unwrap_or_default();
    let back_matter: Vec<String> = ms.and_then(|m| m.get("back_matter_sections")).and_then(|a| a.as_array())
        .map(|arr| arr.iter().filter_map(|v| v.as_str().map(|s| s.to_lowercase().trim().to_string())).collect())
        .unwrap_or_default();

    // Step 3: First pass — resolve canonical names via aliases
    let mut normalized: Vec<Value> = Vec::new();
    for sec in &sections {
        let heading = sec
            .get("heading")
            .and_then(|h| h.as_str())
            .map(|s| s.trim().to_string())
            .unwrap_or_default();
        let heading_lower = heading.to_lowercase();
        let mut canonical: Option<String> = None;

        // 3a. Exact alias match
        for (alias_lower, canon_name) in &alias_to_canonical {
            if heading_lower == *alias_lower {
                canonical = Some(canon_name.clone());
                break;
            }
        }

        // 3b. Substring alias match — longest wins
        if canonical.is_none() && !heading_lower.is_empty() {
            let mut best_len = 0usize;
            let mut best_canon: Option<&str> = None;
            for (alias_lower, canon_name) in &alias_to_canonical {
                if heading_lower.contains(alias_lower.as_str()) && alias_lower.len() > best_len {
                    best_canon = Some(canon_name.as_str());
                    best_len = alias_lower.len();
                }
            }
            canonical = best_canon.map(String::from);
        }

        // 3c. Fall back to section_map's existing name
        if canonical.is_none() {
            canonical = sec.get("name").and_then(|n| n.as_str()).map(String::from);
        }

        let mut sec_val = sec.clone();
        if let Some(obj) = sec_val.as_object_mut() {
            obj.insert(
                "canonical_name".to_string(),
                Value::String(canonical.unwrap_or_default()),
            );
        }
        normalized.push(sec_val);
    }

    // Step 4: Second pass — apply position-based rules
    // Collect overrides first (needs immutable access to normalized for condition checks)
    let mut overrides: Vec<(usize, String)> = Vec::new();
    for (i, sec) in normalized.iter().enumerate() {
        let heading = sec
            .get("heading")
            .and_then(|h| h.as_str())
            .map(|s| s.trim().to_string())
            .unwrap_or_default();
        let heading_lower = heading.to_lowercase();

        let rules_for_heading: Vec<(String, String)> = rule_lookup
            .iter()
            .find(|(k, _)| k == &heading_lower)
            .map(|(_, entries)| entries.clone())
            .unwrap_or_default();

        for (canon, condition) in &rules_for_heading {
            if condition_holds(condition, i, &normalized) {
                overrides.push((i, canon.clone()));
                break;
            }
        }
    }

    // Apply overrides
    for (i, canon) in &overrides {
        if let Some(obj) = normalized[*i].as_object_mut() {
            obj.insert(
                "canonical_name".to_string(),
                Value::String(canon.clone()),
            );
        }
    }

    // Step 5: Classify each section
    let mut classifications: serde_json::Map<String, Value> = serde_json::Map::new();
    for sec in &normalized {
        let canon = sec
            .get("canonical_name")
            .and_then(|c| c.as_str())
            .unwrap_or("");
        let name = sec.get("name").and_then(|n| n.as_str()).unwrap_or("");

        let class = if is_front_matter(canon, &front_matter) {
            "front_matter"
        } else if is_back_matter(canon, &back_matter) {
            "back_matter"
        } else if is_main_text(canon, &main_text_order) {
            "main_text"
        } else {
            "main_text"
        };
        classifications.insert(name.to_string(), Value::String(class.to_string()));
    }

    // Step 6: Build canonical_order (deduplicated)
    let mut canonical_order: Vec<Value> = Vec::new();
    let mut seen = std::collections::HashSet::new();
    for sec in &normalized {
        let cn = sec
            .get("canonical_name")
            .and_then(|c| c.as_str())
            .unwrap_or("");
        if !cn.is_empty() && seen.insert(cn.to_string()) {
            canonical_order.push(Value::String(cn.to_string()));
        }
    }

    serde_json::json!({
        "sections": normalized,
        "canonical_order": canonical_order,
        "classifications": Value::Object(classifications),
    })
}

// ── Condition evaluators ──

fn condition_holds(condition: &str, idx: usize, sections: &[Value]) -> bool {
    if condition == "before_introduction" {
        for prev in sections.iter().take(idx) {
            if prev
                .get("canonical_name")
                .and_then(|c| c.as_str())
                .map_or(false, |cn| cn == "introduction")
            {
                return false;
            }
        }
        return true;
    }

    if condition == "inside_or_immediately_after_discussion" {
        let parent = sections[idx]
            .get("parent_section")
            .and_then(|p| p.as_str());
        if let Some(parent_name) = parent {
            if canonical_of_parent(parent_name, sections)
                .as_deref()
                == Some("discussion")
            {
                return true;
            }
        }
        if idx > 0
            && sections[idx - 1]
                .get("canonical_name")
                .and_then(|c| c.as_str())
                .map_or(false, |cn| cn == "discussion")
        {
            return true;
        }
        return false;
    }

    if condition == "inside_methods" {
        let parent = sections[idx]
            .get("parent_section")
            .and_then(|p| p.as_str());
        if let Some(parent_name) = parent {
            if canonical_of_parent(parent_name, sections).as_deref() == Some("methods") {
                return true;
            }
        }
        return false;
    }

    false
}

fn canonical_of_parent(parent_name: &str, sections: &[Value]) -> Option<String> {
    for sec in sections {
        if sec.get("name").and_then(|n| n.as_str()) == Some(parent_name) {
            return sec
                .get("canonical_name")
                .and_then(|c| c.as_str())
                .map(String::from);
        }
    }
    None
}

// ── Classification helpers ──

fn normalize_for_classification(canonical: &str) -> String {
    let mut s = canonical.to_string();
    for suffix in &["_subsection", "_or_conclusion"] {
        if let Some(pos) = s.find(suffix) {
            s = s[..pos].to_string();
        }
    }
    s.to_lowercase().trim().to_string()
}

fn is_front_matter(canonical: &str, front_list: &[String]) -> bool {
    let base = normalize_for_classification(canonical);
    front_list.iter().any(|f| f.as_str() == base.as_str())
}

fn is_back_matter(canonical: &str, back_list: &[String]) -> bool {
    let base = normalize_for_classification(canonical);
    back_list.iter().any(|b| b.as_str() == base.as_str())
}

fn is_main_text(canonical: &str, main_text_order: &[String]) -> bool {
    let base = normalize_for_classification(canonical);
    main_text_order
        .iter()
        .any(|m| m.as_str() == base.as_str())
}
